src/Data/Primitive/Unlifted/Array/ST.hs

```diff
@@ -0,0 +1,497 @@
+  = Exts.isTrue# (sameMutableUnliftedArray# maa1# maa2#)
+{-# inline sameMutableUnliftedArray #-}
+
+-- | Copies the contents of an immutable array into a mutable array.
+copyUnliftedArray
+  :: MutableUnliftedArray s a -- ^ destination
+  -> Int -- ^ offset into destination
+  -> UnliftedArray a -- ^ source
+  -> Int -- ^ offset into source
+  -> Int -- ^ number of elements to copy
+  -> ST s ()
+{-# inline copyUnliftedArray #-}
+copyUnliftedArray
+  (MutableUnliftedArray dst) (I# doff)
+  (UnliftedArray src) (I# soff) (I# ln) =
+    primitive_ $ copyUnliftedArray# src soff dst doff ln
+
+-- | Copies the contents of one mutable array into another.
+copyMutableUnliftedArray
+  :: MutableUnliftedArray s a -- ^ destination
+  -> Int -- ^ offset into destination
+  -> MutableUnliftedArray s a -- ^ source
+  -> Int -- ^ offset into source
+  -> Int -- ^ number of elements to copy
+  -> ST s ()
+{-# inline copyMutableUnliftedArray #-}
+copyMutableUnliftedArray
+  (MutableUnliftedArray dst) (I# doff)
+  (MutableUnliftedArray src) (I# soff) (I# ln) =
+    primitive_ $ copyMutableUnliftedArray# src soff dst doff ln
+
+-- | Freezes a portion of a 'MutableUnliftedArray', yielding an 'UnliftedArray'.
+-- This operation is safe, in that it copies the frozen portion, and the
+-- existing mutable array may still be used afterward.
+freezeUnliftedArray
+  :: MutableUnliftedArray s a -- ^ source
+  -> Int -- ^ offset
+  -> Int -- ^ length
+  -> ST s (UnliftedArray a)
+freezeUnliftedArray (MutableUnliftedArray mary) (I# off) (I# len) =
+    ST $ \s -> case freezeUnliftedArray# mary off len s of
+      (# s', ary #) -> (# s', UnliftedArray ary #)
+{-# inline freezeUnliftedArray #-}
+
+-- | Thaws a portion of an 'UnliftedArray', yielding a 'MutableUnliftedArray'.
+-- This copies the thawed portion, so mutations will not affect the original
+-- array.
+thawUnliftedArray
+  :: UnliftedArray a -- ^ source
+  -> Int -- ^ offset
+  -> Int -- ^ length
+  -> ST s (MutableUnliftedArray s a)
+{-# inline thawUnliftedArray #-}
+thawUnliftedArray (UnliftedArray ary) (I# off) (I# len) =
+    ST $ \s -> case thawUnliftedArray# ary off len s of
+      (# s', mary #) -> (# s', MutableUnliftedArray mary #)
+
+-- | Execute a stateful computation and freeze the resulting array.
+runUnliftedArray
+  :: (forall s. ST s (MutableUnliftedArray s a))
+  -> UnliftedArray a
+{-# INLINE runUnliftedArray #-}
+-- This is what we'd like to write, but GHC does not yet
+-- produce properly unboxed code when we do
+-- runUnliftedArray m = runST $ m >>= unsafeFreezeUnliftedArray
+runUnliftedArray m = UnliftedArray (runUnliftedArray# m)
+
+runUnliftedArray#
+  :: (forall s. ST s (MutableUnliftedArray s a))
+  -> UnliftedArray# (Unlifted a)
+runUnliftedArray# m = case Exts.runRW# $ \s ->
+  case unST m s of { (# s', MutableUnliftedArray mary# #) ->
+  unsafeFreezeUnliftedArray# mary# s'} of (# _, ary# #) -> ary#
+{-# INLINE runUnliftedArray# #-}
```

Comment:
This could potentially be an expensive action. Do we want to `noDuplicate#` to guarantee it will only be done in one thread? That has its own costs, but it might be worthwhile. *sigh*

Reply:
One option: punt by adding `dupableRunUnliftedArray`.